tutor k8s start fails with no PatchStrategicMerge

<!-- Are you quite sure that you followed the instructions from the Troubleshooting section in the Tutor documentation? https://docs.tutor.overhang.io/troubleshooting.html -->
<!-- If not, please take the time to read them. -->

**Bug description**
<!-- A clear and concise description of what the bug is. Don't be afraid to include a lot of context and as much background information as possible. If applicable, add screenshots to help explain your problem. -->
<!-- Also include relevant error logs; for instance, to debug the LMS provide the output of `tutor local logs lms --tail=100` -->
<!-- Is this a bug in Tutor, or in Open edX? Open edX bugs should not be reported here, but in the Open edX bug tracker: https://openedx.atlassian.net/ -->

We receive a kustomize error while running `tutor k8s start`:

```
Run tutor k8s start
kubectl get namespaces openedx-prod
NAME           STATUS   AGE
openedx-prod   Active   49d
Namespace already exists: skipping creation.
kubectl apply --kustomize /home/runner/.local/share/tutor/env --wait --selector app.kubernetes.io/component=volume
error: trouble configuring builtin PatchStrategicMergeTransformer with config: `
paths:
- k8s/override.yml
`: patch appears to be empty; files=[k8s/override.yml], Patch=
Error: Command failed with status 1: kubectl apply --kustomize /home/runner/.local/share/tutor/env --wait --selector app.kubernetes.io/component=volume
Error: Process completed with exit code 1.
```
We don't specify a `k8s-override` patch, so [tutor/tutor/templates/k8s/override.yml ](https://github.com/overhangio/tutor/blob/master/tutor/templates/k8s/override.yml) results in an empty file.

**How to reproduce**
<!-- Include the FULL, EXACT command that you are running and that is causing an error. Also include the FULL, EXACT command output that you see in the console. THIS IS ESPECIALLY IMPORTANT! Thank you :) -->
<!-- In case of a web error, include the URL of the page that is causing the error.-->
<!-- Did you make any manual changes to the generated environment or config.yml files? If yes, describe those changes.-->

Please se above for most of the details.

I don't believe we have any manual changes to config.yml.

**Environment**
<!-- Include your OS (e.g. Mac OS, Ubuntu 18.04...) and the Tutor version that you use, which can be obtained with `tutor --version`. -->

We're running this in a custom github action running on `ubuntu-latest`, using tutor version 14.0.3

Reverting to 14.0.2 fixed our automation.

**Additional context**
<!-- Add any other context about the problem here.-->

It looks like `patchesStrategicMerge` just needs to be omitted from [kustomization.yml](https://github.com/overhangio/tutor/blob/master/tutor/templates/kustomization.yml) when we're not customizing any base resources.

This is based on peeking at kustomize's [PatchStrategicMergeTransformer](https://github.com/kubernetes-sigs/kustomize/blob/master/plugin/builtin/patchstrategicmergetransformer/PatchStrategicMergeTransformer.go#L50)

--- a/changelog.d/20230214_105510_keith_fix_jobs_merge.md
+++ b/changelog.d/20230214_105510_keith_fix_jobs_merge.md
@@ -1,0 +1,12 @@
+<!--
+Create a changelog entry for every new user-facing change. Please respect the following instructions:
+- Indicate breaking changes by prepending an explosion 💥 character.
+- Prefix your changes with either [Bugfix], [Improvement], [Feature], [Security], [Deprecation].
+- You may optionally append "(by @<author>)" at the end of the line, where "<author>" is either one (just one)
+  of your GitHub username, real name or affiliated organization. These affiliations will be displayed in
+  the release notes for every release.
+-->
+
+<!-- - 💥[Feature] Foobarize the blorginator. This breaks plugins by renaming the `FOO_DO` filter to `BAR_DO`. (by @regisb) -->
+<!-- - [Improvement] This is a non-breaking change. Life is good. (by @billgates) -->
+[Bugfix] `patchStrategicMerge` can now be applied to jobs (by @keithgg)

--- a/tutor/commands/k8s.py
+++ b/tutor/commands/k8s.py
@@ -65,7 +65,8 @@ class K8sTaskRunner(BaseTaskRunner):
 
     def run_task(self, service: str, command: str) -> int:
         canonical_job_name = f"{service}-job"
-        job = self.load_job(canonical_job_name)
+        all_jobs = list(self._load_jobs())
+        job = self._find_job(canonical_job_name, all_jobs)
         # Create a unique job name to make it deduplicate jobs and make it easier to
         # find later. Logs of older jobs will remain available for some time.
         job_name = canonical_job_name + "-" + datetime.now().strftime("%Y%m%d%H%M%S")
@@ -99,7 +100,10 @@ class K8sTaskRunner(BaseTaskRunner):
         job["spec"]["backoffLimit"] = 1
         job["spec"]["ttlSecondsAfterFinished"] = 3600
 
-        self._replace_job(canonical_job_name, job)
+        with open(
+            tutor_env.pathjoin(self.root, "k8s", "jobs.yml"), "w", encoding="utf-8"
+        ) as job_file:
+            serialize.dump_all(all_jobs, job_file)
 
         # We cannot use the k8s API to create the job: configMap and volume names need
         # to be found with the right suffixes.
@@ -141,22 +145,22 @@ class K8sTaskRunner(BaseTaskRunner):
         """
         Find a given job definition in the rendered k8s/jobs.yml template.
         """
-        return self._extract_job(name, self._load_jobs())[1]
+        return self._find_job(name, self._load_jobs())
 
-    def _extract_job(self, name: str, all_jobs: Iterable[Any]) -> tuple[int, Any]:
+    def _find_job(self, name: str, all_jobs: Iterable[Any]) -> Any:
         """
         Find the matching job definition in the in the list of jobs provided.
 
-        Returns a tuple with the index of the found job and it's manifest.
+        Returns the found job's manifest.
         """
-        for index, job in enumerate(all_jobs):
+        for job in all_jobs:
             job_name = job["metadata"]["name"]
             if not isinstance(job_name, str):
                 raise exceptions.TutorError(
                     f"Invalid job name: '{job_name}'. Expected str."
                 )
             if job_name == name:
-                return index, job
+                return job
         raise exceptions.TutorError(f"Could not find job '{name}'")
 
     def _load_jobs(self) -> Iterable[Any]:
@@ -164,20 +168,6 @@ class K8sTaskRunner(BaseTaskRunner):
         for manifest in serialize.load_all(manifests):
             if manifest["kind"] == "Job":
                 yield manifest
-
-    def _replace_job(self, job_name: str, job: Config) -> List[Config]:
-        """
-        Replaces the job matching job_name in jobs.yml with the specified job.
-        """
-        all_jobs = list(self._load_jobs())
-        job_index = self._extract_job(job_name, all_jobs)[0]
-        all_jobs[job_index] = job
-        with open(
-            tutor_env.pathjoin(self.root, "k8s", "jobs.yml"), "w", encoding="utf-8"
-        ) as job_file:
-            serialize.dump_all(all_jobs, job_file)
-
-        return all_jobs
 
     def active_job_names(self) -> List[str]:
         """

--- a/tutor/commands/k8s.py
+++ b/tutor/commands/k8s.py
@@ -1,6 +1,6 @@
 from datetime import datetime
 from time import sleep
-from typing import Any, List, Optional, Type
+from typing import Any, List, Optional, Type, Iterable
 
 import click
 
@@ -64,11 +64,11 @@ class K8sTaskRunner(BaseTaskRunner):
     """
 
     def run_task(self, service: str, command: str) -> int:
-        job_name = f"{service}-job"
-        job = self.load_job(job_name)
+        canonical_job_name = f"{service}-job"
+        job = self.load_job(canonical_job_name)
         # Create a unique job name to make it deduplicate jobs and make it easier to
         # find later. Logs of older jobs will remain available for some time.
-        job_name += "-" + datetime.now().strftime("%Y%m%d%H%M%S")
+        job_name = canonical_job_name + "-" + datetime.now().strftime("%Y%m%d%H%M%S")
 
         # Wait until all other jobs are completed
         while True:
@@ -98,11 +98,9 @@ class K8sTaskRunner(BaseTaskRunner):
         job["spec"]["template"]["spec"]["containers"][0]["args"] = container_args
         job["spec"]["backoffLimit"] = 1
         job["spec"]["ttlSecondsAfterFinished"] = 3600
-        # Save patched job to "jobs.yml" file
-        with open(
-            tutor_env.pathjoin(self.root, "k8s", "jobs.yml"), "w", encoding="utf-8"
-        ) as job_file:
-            serialize.dump(job, job_file)
+
+        self._replace_job(canonical_job_name, job)
+
         # We cannot use the k8s API to create the job: configMap and volume names need
         # to be found with the right suffixes.
         kubectl_apply(
@@ -143,16 +141,43 @@ class K8sTaskRunner(BaseTaskRunner):
         """
         Find a given job definition in the rendered k8s/jobs.yml template.
         """
-        all_jobs = self.render("k8s", "jobs.yml")
-        for job in serialize.load_all(all_jobs):
+        return self._extract_job(name, self._load_jobs())[1]
+
+    def _extract_job(self, name: str, all_jobs: Iterable[Any]) -> tuple[int, Any]:
+        """
+        Find the matching job definition in the in the list of jobs provided.
+
+        Returns a tuple with the index of the found job and it's manifest.
+        """
+        for index, job in enumerate(all_jobs):
             job_name = job["metadata"]["name"]
             if not isinstance(job_name, str):
                 raise exceptions.TutorError(
                     f"Invalid job name: '{job_name}'. Expected str."
                 )
             if job_name == name:
-                return job
+                return index, job
         raise exceptions.TutorError(f"Could not find job '{name}'")
+
+    def _load_jobs(self) -> Iterable[Any]:
+        manifests = self.render("k8s", "jobs.yml")
+        for manifest in serialize.load_all(manifests):
+            if manifest["kind"] == "Job":
+                yield manifest
+
+    def _replace_job(self, job_name: str, job: Config) -> List[Config]:
+        """
+        Replaces the job matching job_name in jobs.yml with the specified job.
+        """
+        all_jobs = list(self._load_jobs())
+        job_index = self._extract_job(job_name, all_jobs)[0]
+        all_jobs[job_index] = job
+        with open(
+            tutor_env.pathjoin(self.root, "k8s", "jobs.yml"), "w", encoding="utf-8"
+        ) as job_file:
+            serialize.dump_all(all_jobs, job_file)
+
+        return all_jobs
 
     def active_job_names(self) -> List[str]:
         """

--- a/tutor/serialize.py
+++ b/tutor/serialize.py
@@ -17,6 +17,10 @@ def load_all(stream: str) -> t.Iterator[t.Any]:
     return yaml.load_all(stream, Loader=yaml.SafeLoader)
 
 
+def dump_all(content: t.Any, fileobj: TextIOWrapper) -> None:
+    yaml.safe_dump_all(content, stream=fileobj, default_flow_style=False)
+
+
 def dump(content: t.Any, fileobj: TextIOWrapper) -> None:
     yaml.dump(content, stream=fileobj, default_flow_style=False)
 

--- a/tutor/serialize.py
+++ b/tutor/serialize.py
@@ -17,8 +17,8 @@ def load_all(stream: str) -> t.Iterator[t.Any]:
     return yaml.load_all(stream, Loader=yaml.SafeLoader)
 
 
-def dump_all(content: t.Any, fileobj: TextIOWrapper) -> None:
-    yaml.safe_dump_all(content, stream=fileobj, default_flow_style=False)
+def dump_all(documents: t.Sequence[t.Any], fileobj: TextIOWrapper) -> None:
+    yaml.safe_dump_all(documents, stream=fileobj, default_flow_style=False)
 
 
 def dump(content: t.Any, fileobj: TextIOWrapper) -> None:

